core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/federation/JoinExecutorBase.java

```diff
@@ -24,20 +25,29 @@
  */
 public abstract class JoinExecutorBase<T> extends LookAheadIteration<T, QueryEvaluationException> {
 
+	/**
+	 * @deprecated No replacement, don't use static shared int variables.
+	 */
 	protected static int NEXT_JOIN_ID = 1;
 
 	/* Constants */
 	protected final TupleExpr rightArg; // the right argument for the join
 
 	protected final BindingSet bindings; // the bindings
 
-	protected CloseableIteration<T, QueryEvaluationException> leftIter;
+	protected final CloseableIteration<T, QueryEvaluationException> leftIter;
 
-	protected CloseableIteration<T, QueryEvaluationException> rightIter;
+	protected volatile CloseableIteration<T, QueryEvaluationException> rightIter;
 
-	protected volatile boolean closed;
+	/**
+	 * @deprecated Use {@link AbstractCloseableIteration#isClosed()} instead.
+	 */
+	protected volatile boolean closed = false;
```

Comment:
This is still set to true in the close method incase third-party code is relying on it, but is not accessed anymore so could be removed in future.